src/sync.rs

```diff
@@ -96,12 +96,26 @@ struct Condvar {
     waiters: VecDeque<CondvarWaiter>,
 }
 
+/// The futex state.
+#[derive(Default, Debug)]
+struct Futex {
+    waiters: VecDeque<FutexWaiter>,
+}
+
+/// A thread waiting on a futex.
+#[derive(Debug)]
+struct FutexWaiter {
+    /// The thread that is waiting on this futex.
+    thread: ThreadId,
+}
+
 /// The state of all synchronization variables.
 #[derive(Default, Debug)]
 pub(super) struct SynchronizationState {
     mutexes: IndexVec<MutexId, Mutex>,
     rwlocks: IndexVec<RwLockId, RwLock>,
     condvars: IndexVec<CondvarId, Condvar>,
+    futexes: HashMap<Pointer<stacked_borrows::Tag>, Futex>,
```

Comment:
Using `Pointer<stacked_borrows::Tag>` here seems odd, it means the same pointer with different tags is tracked separately.

I think it makes more sense to just use `Pointer`. You can use `erase_tag` to get that from a tagged pointer.

Reply:
Ah, I should've probably read a bit about what a pointer tag means. :)  It's erased now.